Revise Pod concept (#22603)
kubernetes/website: content/en/docs/concepts/workloads/pods/_index.md

---
reviewers:
- erictune
title: Pods
content_type: concept
weight: 10
no_list: true
card:
  name: concepts
  weight: 60
---

<!-- overview -->

> _Pods_ are the smallest deployable units of computing that can be created and
> managed in Kubernetes.

Reviewer: ```suggestion
_Pods_ are the smallest deployable units of computing that you can create and manage in Kubernetes.
```

> A _Pod_ (as in a pod of whales or pea pod) is a group of one or more
> {{< glossary_tooltip text="containers" term_id="container" >}} (such as
> Docker containers), with shared storage/network, and a specification
> for how to run the containers.  A Pod's contents are always co-located and
> co-scheduled, and run in a shared context.  A Pod models an
> application-specific "logical host"&mdash;it contains one or more application
> containers which are relatively tightly coupled&mdash;in a pre-container
> world, being executed on the same physical or virtual machine would mean being
> executed on the same logical host.

Reviewer: ```suggestion
A _Pod_ (as in a pod of whales or pea pod) is a group of one or more
{{< glossary_tooltip text="containers" term_id="container" >}}, with shared storage/network resources, and a specification
for how to run the containers. A Pod's contents are always co-located and
co-scheduled, and run in a shared context. A Pod models an
application-specific "logical host": it contains one or more application
containers which are relatively tightly coupled. 
In non-cloud contexts, applications executed on the same physical or virtual machine are analogous to cloud applications executed on the same logical host.
```

I think I got the gist of what you were trying to say here, but I might be off base. Feel free to ignore all/parts of this suggestion.

Author: These aren't actually my words. Happy to tweak them though.

As well as application containers, a Pod can contain
[init containers](/docs/concepts/workloads/pods/init-containers/) that run
during Pod startup. You can also inject
[ephemeral containers](/docs/concepts/workloads/pods/ephemeral-containers/)
for debugging if your cluster offers this.

<!-- body -->

## What is a Pod?

{{< note >}}
While Kubernetes supports more
{{< glossary_tooltip text="container runtimes" term_id="container-runtime" >}}
than just Docker, [Docker](https://www.docker.com/) is the most commonly known
runtime, and it helps to describe Pods using some terminology from Docker.
{{< /note >}}

The shared context of a Pod is a set of Linux namespaces, cgroups, and
potentially other facets of isolation - the same things that isolate a Docker
container.  Within a Pod's context, the individual applications may have
further sub-isolations applied.

In terms of Docker concepts, a Pod is similar to a group of Docker containers
with shared namespaces and shared filesystem volumes.

## Using Pods

> Usually you dont' need to create Pods directly. Instead, you can almost
> always use workload resources such as {{< glossary_tooltip text="Deployment"
> term_id="deployment" >}} or {{< glossary_tooltip text="Job" term_id="job" >}},
> even for singleton Pods.

Reviewer: ```suggestion
Usually you don't need to create Pods directly, even singleton Pods. Instead, create them using workload resources such as {{< glossary_tooltip text="Deployment"
term_id="deployment" >}} or {{< glossary_tooltip text="Job" term_id="job" >}}.
```

> If your Pods need to track state, there is a
> {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}} resource intended
> to be the best fit for that use case.

Reviewer: ```suggestion
If your Pods need to track state, use the
{{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}} resource.
```

Author: StatefulSet isn't always the best choice for every case that involves stateful apps; I wanted to capture that nuance.

Suggestions?

Reviewer: ```suggestion
If your Pods need to track state, consider the 
{{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}} resource.
```

> Pods in a Kubernetes cluster can be used in two main ways:

Reviewer: ```suggestion
Pods in a Kubernetes cluster are used in two main ways:
```

> * **Pods that run a single container**. The "one-container-per-Pod" model is the
>   most common Kubernetes use case; in this case, you can think of a Pod as a
>   wrapper around a single container; Kubernetes manages Pods rather than managing
>   the containers directly.

Reviewer: ```suggestion
* **Pods that run a single container**. The "one-container-per-Pod" model is the
  most common Kubernetes use case. Think of a Pod as a
  wrapper around a single container; Kubernetes manages the Pod instead of managing
  the containers directly.
```

> * **Pods that run multiple containers that need to work together**. A Pod might
>   encapsulate an application composed of multiple co-located containers that are
>   tightly coupled and need to share resources. These co-located containers might
>   form a single cohesive unit of service—one container serving data stored in a
>   shared volume to the public, while a separate _sidecar_ container refreshes or
>   updates those files.  
>   The Pod wraps these containers, storage resources, and an ephemeral network
>   identity together as a single unit.

Reviewer: ```suggestion
* **Pods that run multiple containers that need to work together**. A Pod can
  encapsulate an application composed of multiple co-located containers that are
  tightly coupled and need to share resources. These co-located containers
  form a single cohesive unit of service—one container serving data stored in a
  shared volume to the public, while a separate _sidecar_ container refreshes or
  updates those files.  
  The Pod wraps these containers, storage resources, and an ephemeral network
  identity together as a single unit.
```

Author: I'll go with:
```suggestion
* **Pods that run multiple containers that need to work together**. A Pod can
  encapsulate an application composed of multiple co-located containers that are
  tightly coupled and need to share resources. These co-located containers
  form a single cohesive unit of service—for example, one container serving data
  stored in a shared volume to the public, while a separate _sidecar_ container
  refreshes or updates those files.  
  The Pod wraps these containers, storage resources, and an ephemeral network
  identity together as a single unit.
```

  {{< note >}}
  Grouping multiple co-located and co-managed containers in a single Pod is a
  relatively advanced use case. You should use this pattern only in specific
  instances in which your containers are tightly coupled.
  {{< /note >}}

Each Pod is meant to run a single instance of a given application. If you want to
scale your application horizontally (to provide more overall resources by running
more instances), you should use multiple Pods, one for each instance. In
Kubernetes, this is typically referred to as _replication_.
Replicated Pods are usually created and managed as a group by a workload resource
and its {{< glossary_tooltip text="controller" term_id="controller" >}}.

See [Pods and controllers](#pods-and-controllers) for more information on how
Kubernetes uses workload resources, and their controllers, to implement application
scaling and auto-healing.

### How Pods manage multiple containers

Pods are designed to support multiple cooperating processes (as containers) that form
a cohesive unit of service. The containers in a Pod are automatically co-located and
co-scheduled on the same physical or virtual machine in the cluster. The containers

> can share resources and dependencies, communicate with one another, and coordinate

Reviewer: ```suggestion
share resources and dependencies, communicate with one another, and coordinate
```

Author: I think the original text is more accurate. “These containers coordinate when and how they are terminate” is usually _not_ true.

### Pods and controllers

You can use workload resources to create and manage multiple Pods for you. A controller
for the resource handles replication and rollout and automatic healing in case of
Pod failure. For example, if a Node fails, a controller notices that Pods on that
Node have stopped working and creates a replacement Pod. The scheduler places the
replacement Pod onto a healthy Node.

Here are some examples of workload resources that manage one or more Pods:

* {{< glossary_tooltip text="Deployment" term_id="deployment" >}}
* {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
* {{< glossary_tooltip text="DaemonSet" term_id="daemonset" >}}

### Pod templates

Controllers for {{< glossary_tooltip text="workload" term_id="workload" >}} resources create Pods
from a _pod template_ and manage those Pods on your behalf.

PodTemplates are specifications for creating Pods, and are included in workload resources such as
[Deployments](/docs/concepts/workloads/controllers/deployment/),
[Jobs](/docs/concepts/jobs/run-to-completion-finite-workloads/), and
[DaemonSets](/docs/concepts/workloads/controllers/daemonset/).

> Each controller for a workload resource uses the `PodTemplate inside the workload

Reviewer: ```suggestion
Each controller for a workload resource uses the `PodTemplate` inside the workload
```

> object to make actual Pods. The PodTemplate is part of the desired state of whatever

Reviewer: ```suggestion
object to make actual Pods. The `PodTemplate` is part of the desired state of whatever
```

workload resource you used to run your app.

The sample below is a manifest for a simple Job with a `template` that starts one
container. The container in that Pod prints a message then pauses.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: hello
spec:
  template:
    # This is the pod template
    spec:
      containers:
      - name: hello
        image: busybox
        command: ['sh', '-c', 'echo "Hello, Kubernetes!" && sleep 3600']
      restartPolicy: OnFailure
    # The pod template ends here
```

Modifying the pod template or switching to a new pod template has no effect on the

> Pods that already exist. Pods do not receive template updates directly; instead,

Reviewer: ```suggestion
Pods that already exist. Pods do not receive template updates directly. Instead,
```

This means that the Pods running on a node are visible on the API server,
but cannot be controlled from there.

## {{% heading "whatsnext" %}}

* Learn about the [lifecycle of a Pod](/docs/concepts/workloads/pods/pod-lifecycle/).
* Learn about [PodPresets](/docs/concepts/workloads/pods/podpreset/).
* Lean about [RuntimeClass](/docs/concepts/containers/runtime-class/) and how you can use it to
  configure different Pods with different container runtime configurations.
* Read about [Pod topology spread constraints](/docs/concepts/workloads/pods/pod-topology-spread-constraints/).
* Read about [PodDisruptionBudget](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) and how you can use it to manage application availability during disruptions.
* Pod is a top-level resource in the Kubernetes REST API.  
  The [Pod](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#pod-v1-core)
  object definition describes the object in detail.
* [The Distributed System Toolkit: Patterns for Composite Containers](https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns) explains common layouts for Pods with more than one container.

> To understand the context for why Kubernetes wraps a common Pod API in other resources (such as {{< glossary_tooltip text="StatefulSets" term_id="statefulset" >}} or {{< glossary_tooltip text="Deployments" term_id="deployment" >}}, you can read about the prior art, including:

Reviewer: Prior... art..?

Author: Is that wording unhelpful? It's accurate, but maybe it's obscure.
